libyul/AsmParser.cpp

```diff
@@ -241,10 +263,21 @@ Case Parser::parseCase()
 	return _case;
 }
 
+template <typename Incrementable>
+auto scopedIncrement(Incrementable& incrementable)
```

Comment:
Hm, would it look less complicated to just define a class that does everything instead of wrapping it inside a function?
Also, this could go into devcore.

Reply:
sure, I thought about `libdevcore`, too. on the other concern, I believe I did code a bit too much F# lately. :-)

Reply:
I could rewrite it into a struct with the use of deduction guides:

```cpp
template<typename U>
struct ScopedIncrement {
        U& value;
        explicit ScopedIncrement(U& _value) : value{ ++_value } {}
        ~ScopedIncrement() { --value; }
};

template <class T>
ScopedIncrement(T& _value) -> ScopedIncrement<T>;
```

So that the use of it actually looks a simple as this:
```cpp
auto guard = scopedIncrement(m_myIncrementableValue);
```
But that's C++17, and I wanted to avoid requiring the caller to explicitly state the value type (too verbose, could get guessed wrong).

Reply:
```
template <typename T>
struct ScopedIncrement {
        T& value;
        explicit ScopedIncrement(T& _value) : value{ ++_value } {}
        ~ScopedIncrement() { --value; }
};
```

```
ScopedIncrement increment{m_myIncrementableValue};
```

 - ah the problem is that you have to explicitly provide the template argument for classes, but not for functions, right?

Reply:
exactly. (unless you can use deduction guides, which are introduced in C++17. I'm not up-to-date how the status is on that intend. (don't wanna force-push here for it anyways :) ).